internal/collector/prometheus/prometheus_collector.go

```diff
@@ -0,0 +1,847 @@
+		constants.LabelNamespace, deployNamespace,
+		constants.VLLMTimeToFirstTokenSecondsCount,
+		constants.LabelModelName, modelName,
+		constants.LabelNamespace, deployNamespace)
+
+	// Metric 5: Average ITL (Inter-Token Latency) - Prometheus returns seconds, converted to msec
+	itlQuery := fmt.Sprintf(`sum(rate(%s{%s="%s",%s="%s"}[1m]))/sum(rate(%s{%s="%s",%s="%s"}[1m]))`,
+		constants.VLLMTimePerOutputTokenSecondsSum,
+		constants.LabelModelName, modelName,
+		constants.LabelNamespace, deployNamespace,
+		constants.VLLMTimePerOutputTokenSecondsCount,
+		constants.LabelModelName, modelName,
+		constants.LabelNamespace, deployNamespace)
+
+	// --- 2. Execute Queries ---
+
+	arrivalVal, err := pc.queryAndExtractMetric(ctx, arrivalQuery, "ArrivalRate")
+	if err != nil {
+		return llmdVariantAutoscalingV1alpha1.Allocation{}, err
+	}
+	arrivalVal *= 60 // convert from req/sec to req/min
+
+	avgInputTokens, err := pc.queryAndExtractMetric(ctx, avgPromptToksQuery, "AvgInputTokens")
+	if err != nil {
+		return llmdVariantAutoscalingV1alpha1.Allocation{}, err
+	}
+
+	avgOutputTokens, err := pc.queryAndExtractMetric(ctx, avgDecToksQuery, "AvgOutputTokens")
+	if err != nil {
+		return llmdVariantAutoscalingV1alpha1.Allocation{}, err
+	}
+
+	ttftAverageTime, err := pc.queryAndExtractMetric(ctx, ttftQuery, "TTFTAverageTime")
+	if err != nil {
+		return llmdVariantAutoscalingV1alpha1.Allocation{}, err
+	}
+	// Convert from seconds (Prometheus format) to milliseconds (optimizer format)
+	ttftAverageTime *= 1000
+
+	itlAverage, err := pc.queryAndExtractMetric(ctx, itlQuery, "ITLAverage")
+	if err != nil {
+		return llmdVariantAutoscalingV1alpha1.Allocation{}, err
+	}
+	// Convert from seconds (Prometheus format) to milliseconds (optimizer format)
+	itlAverage *= 1000
+
+	// --- 3. Collect K8s and Static Info ---
+
+	// number of replicas
+	numReplicas := int(*deployment.Spec.Replicas)
+
+	// accelerator type - strict validation required
+	acc := ""
+	if val, ok := va.Labels["inference.optimization/acceleratorName"]; ok && val != "" {
+		acc = val
+	} else {
+		return llmdVariantAutoscalingV1alpha1.Allocation{},
+			fmt.Errorf("missing or empty acceleratorName label on VariantAutoscaling object: %s", va.Name)
+	}
+
+	// cost
+	discoveredCost := float64(*deployment.Spec.Replicas) * acceleratorCostVal
+
+	// max batch size
+	// TODO: collect value from server
+	maxBatch := 256
+
+	// --- 4. Build Allocation ---
+
+	// Format metric values, ensuring they meet CRD validation regex '^\\d+(\\.\\d+)?$'
+	variantCostStr := strconv.FormatFloat(discoveredCost, 'f', 2, 64)
+	ttftAverageStr := strconv.FormatFloat(ttftAverageTime, 'f', 2, 64)
+	itlAverageStr := strconv.FormatFloat(itlAverage, 'f', 2, 64)
+	arrivalRateStr := strconv.FormatFloat(arrivalVal, 'f', 2, 64)
+	avgInputTokensStr := strconv.FormatFloat(avgInputTokens, 'f', 2, 64)
+	avgOutputTokensStr := strconv.FormatFloat(avgOutputTokens, 'f', 2, 64)
+
+	currentAlloc := llmdVariantAutoscalingV1alpha1.Allocation{
```

Comment:
The VA reconciler is responsible for assembling this struct. I would extract this code out and move it under the `controller` pkg.